# Анализ данных на Python
## Практика. Фильтрация и сортировка
#### <i>Малкова Ксения, Преподаватель ФКН НИУ ВШЭ, kemalkova@hse.ru</i>

Анализируем датасэт с информацией об атлетах участвовавших в Олимпийских играх (`olymp.csv`). Скачайте его. Разделитель запятая.

Данные содержат следующие признаки:
- `ID` - уникальный номер спортсмена
- `Name` - имя атлета
- `Sex` - пол (мужской - M,  женский - F)
- `Age` - возраст атлета
- `Height` - рост (в см)
- `Weight` - вес (в кг)
- `Team` - cтрана
- `NOC` - код страны от Олимпийского комитета (3 буквы)
- `Games` - год и сезон Олимпийских игр
- `Year` - год Олимпийских игр
- `Season` - сезон (Зимние - Winter,  Лето - Summer)
- `City` - город проведения Олимпийских игр
- `Sport` - спортивная дисциплина
- `Event` - категория спортивной дисциплины
- `Medal` - достоинство медали (Золото - Gold, Серебро - Silver, Бронза - Bronze, Без медали - NaN)


In [2]:
import pandas as pd
df = pd.read_csv('../../data/olymp.csv')

In [3]:
df.head()

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal  
0       Basketball Men's Basketball   NaN  
1      Judo Men's Extra-Lightweight   NaN  
2           Football Men's Football   NaN  
3       Tug-Of-War Men's Tug-Of-War  Gold  
4  Speed Skating Women's 500 metres   NaN

## Задание 1
Сколько лет было самому молодому участнику олимпийских игр в 1992 году? В ответ запишите целое число

**Алгоритм решения**:
1. Отфильтруйте данные по столбцу `Year` и оставьте только 1992 год
2. Отфильтрованные данные отсортируйте по возрастанию по столбцу `Age`

In [6]:
df[df['Year'] == 1992].sort_values('Age').head(5)

ID                             Name Sex   Age  Height  Weight  \
73461   37333  Carlos Bienvenido Front Barrera   M  11.0     NaN     NaN   
120234  60854                       Judit Kiss   F  12.0   171.0    57.0   
120233  60854                       Judit Kiss   F  12.0   171.0    57.0   
80831   41040       Gina Elena Gogean (-Groza)   F  13.0   150.0    40.0   
80829   41040       Gina Elena Gogean (-Groza)   F  13.0   150.0    40.0   

           Team  NOC        Games  Year  Season       City       Sport  \
73461     Spain  ESP  1992 Summer  1992  Summer  Barcelona      Rowing   
120234  Hungary  HUN  1992 Summer  1992  Summer  Barcelona    Swimming   
120233  Hungary  HUN  1992 Summer  1992  Summer  Barcelona    Swimming   
80831   Romania  ROU  1992 Summer  1992  Summer  Barcelona  Gymnastics   
80829   Romania  ROU  1992 Summer  1992  Summer  Barcelona  Gymnastics   

                                        Event Medal  
73461               Rowing Men's Coxed Eights   NaN  
120234  Swimming Women's 800 metres Freestyle   NaN  
120233  Swimming Women's 400 metres Freestyle   NaN  
80831         Gymnastics Women's Balance Beam   NaN  
80829          Gymnastics Women's Horse Vault   NaN

**Ответ:** 11

## Задание 2
Сколько всего **уникальных** участников Олимпийских игр в 2012 году? В ответ впишите целое число

**Алгоритм решения**:
1. Отфильтруйте данные по признаку `Year` и оставьте только 2012 год
2. К отфильтрованным данным примените метод `.drop_duplicates("признак")`по признаку `ID`, чтобы оставить строки, только с уникальными атлетами
3. Затем примените свойство `.shape` чтобы узнать количество уникальных спортсменов в 2012 году

In [9]:
df[df['Year']==2012].drop_duplicates('ID').shape[0]

10517

**Ответ:** 10517

## Задание 3
Сколько мужчин баскетболистов в принимали участие в Олимпиаде 2012 года? В ответ впишите целое число

**Алгоритм решения**:
1. Отфильтруйте данные по трем условиям: по признаку `Year` (только 2012 год), по признаку `Sex` (только мужчин, **M**) и признаку `Sport` (только баскетболистов, **Basketball**)
2. К отфильтрованным данным примените метод `.drop_duplicates("признак")`по признаку `ID`, чтобы оставить строки, только с уникальными атлетами
3. Затем примените свойство `.shape` чтобы узнать количество мужчин баскетболистов в 2012 году

In [13]:
df[(df['Sex']=='M') & (df['Sport']=='Basketball') & (df['Year']==2012)].drop_duplicates('ID').shape[0]

144

**Ответ:** 144

## Задание 4
Какую долю составляют мужчины-баскетболисты среди всех участников игр 2012 года? Округлите ответ до двух знаков, разделителем является точка.

In [15]:
df_filt_2012_m_bask = df[(df['Sex']=='M') & (df['Sport']=='Basketball') & (df['Year']==2012)].drop_duplicates('ID')
df_filt_2012 = df[df['Year']==2012].drop_duplicates('ID')

df_filt_2012_m_bask.shape[0] / df_filt_2012.shape[0]

0.013692117524008748

**Ответ:** 0.01

## Задание 5
В каких годах проводились соревнования по керлингу (`Curling`)? Ответ: отсортированный по возрастанию список

**Алгоритм решения**:
1. Сделайте фильтрацию и оставьте только соревнования (столбец `Sport`) по керлингу (`Curling`)
2. В отфильтрованном датафрейме обратитесь к столбцу `Year`
3. Примените метод `.unique()` Он вернет список уникальных значений из столбца.
4. Чтобы года были по порядку примените функцию `sorted()`

In [24]:
df[df['Sport']=='Curling'].drop_duplicates('Year').sort_values('Year')['Year'].values

array([1924, 1998, 2002, 2006, 2010, 2014])

**Ответ:** [1924, 1998, 2002, 2006, 2010, 2014]

## Задание 6
В каком году женщины смогли впервые принять участие в Олимпиаде? В ответ впишите целое число

**Алгоритм решения**:
1. Отфильтруйте данные по столбцу `Sex` и оставьте только женщин (**F**)
2. Отфильтрованные данные отсортируйте по возрастанию по столбцу `Year`

In [25]:
df[df['Sex']=='F'].sort_values('Year').head()

ID                                               Name Sex   Age  \
283       150                      Margaret Ives Abbott (-Dunne)   F  23.0   
284       151                Mary Perkins Ives Abbott (Perkins-)   F  42.0   
73432   37319                               Mme. Froment-Meurice   F   NaN   
69404   35364  Jeanne Marie Henriette Filleaul-Brohy (Hantjens-)   F  33.0   
164478  82587                                             Moulin   F   NaN   

        Height  Weight           Team  NOC        Games  Year  Season   City  \
283        NaN     NaN  United States  USA  1900 Summer  1900  Summer  Paris   
284        NaN     NaN  United States  USA  1900 Summer  1900  Summer  Paris   
73432      NaN     NaN         France  FRA  1900 Summer  1900  Summer  Paris   
69404      NaN     NaN         France  FRA  1900 Summer  1900  Summer  Paris   
164478     NaN     NaN         France  FRA  1900 Summer  1900  Summer  Paris   

                Sport                                          Event Medal  
283              Golf                        Golf Women's Individual  Gold  
284              Golf                        Golf Women's Individual   NaN  
73432            Golf                        Golf Women's Individual   NaN  
69404         Croquet               Croquet Mixed Singles, Two Balls   NaN  
164478  Equestrianism  Equestrianism Mixed Hacks And Hunter Combined   NaN

**Ответ:** 1900

## Задание 7
Сколько спортсменок впервые приняли участие в Олимпиаде? В ответ впишите целое число.

**Алгоритм решения**:
1. Отфильтруйте данные и найдите только спортсменок (по столбцу `Sex`), которые приняли участие в первой Олимпиаде, до которой были допущены женщины (по столбцу `Year`) 
2. Удалите повторяющихся спортсменок с помощью метода `.drop_duplicates("ID")`
3. Узнайте количество спортсменок с помощью свойства `.shape`

In [26]:
df[(df['Sex']=='F') & (df['Year']==1900)].drop_duplicates('ID').shape[0]

23

**Ответ:** 23

## Задание 8
В каких видах спорта женщины принимали участие, на первой Олимпиаде до которой они были допущены? Ответ: список с уникальными видами спорта

**Алгоритм решения**:
1. Отфильтруйте данные и найдите только спортсменок (по столбцу `Sex`), которые приняли участие в первой Олимпиаде до которой были допущены женщины (по столбцу `Year`) 
2. Удалите дубликаты по столбцу `Sport`
3. Обратитесь к столбцу Sport и получите его значения (метод `.values`)

In [31]:
df[(df['Sex']=='F') & (df['Year']==1900)].drop_duplicates("Sport")['Sport'].values

array(['Golf', 'Tennis', 'Sailing', 'Croquet', 'Equestrianism'],
      dtype=object)

**Ответ:** ['Golf', 'Tennis', 'Sailing', 'Croquet', 'Equestrianism']

## Задание 9
Определите медаль какого достоинства Усейн Болт (**Usain St. Leo Bolt**) завоевал на своей первой Олимпиаде?

- Серебро
- Бронза
- Золото
- Без медали

**Алгоритм решения**:
1. Сделайте фильтрацию и найдите все строки про Усейна Болта (**Usain St. Leo Bolt**) по столбцу `Name`
2. Отфильтрованные данные отсортируйте по возрастанию по столбцу `Year`
3. Найдите какую медаль он завоевал

In [32]:
df[df['Name']=='Usain St. Leo Bolt'].sort_values('Year')

ID                Name Sex   Age  Height  Weight     Team  NOC  \
24876  13029  Usain St. Leo Bolt   M  17.0   196.0    95.0  Jamaica  JAM   
24877  13029  Usain St. Leo Bolt   M  21.0   196.0    95.0  Jamaica  JAM   
24878  13029  Usain St. Leo Bolt   M  21.0   196.0    95.0  Jamaica  JAM   
24879  13029  Usain St. Leo Bolt   M  21.0   196.0    95.0  Jamaica  JAM   
24880  13029  Usain St. Leo Bolt   M  25.0   196.0    95.0  Jamaica  JAM   
24881  13029  Usain St. Leo Bolt   M  25.0   196.0    95.0  Jamaica  JAM   
24882  13029  Usain St. Leo Bolt   M  25.0   196.0    95.0  Jamaica  JAM   
24883  13029  Usain St. Leo Bolt   M  29.0   196.0    95.0  Jamaica  JAM   
24884  13029  Usain St. Leo Bolt   M  29.0   196.0    95.0  Jamaica  JAM   
24885  13029  Usain St. Leo Bolt   M  29.0   196.0    95.0  Jamaica  JAM   

             Games  Year  Season            City      Sport  \
24876  2004 Summer  2004  Summer          Athina  Athletics   
24877  2008 Summer  2008  Summer         Beijing  Athletics   
24878  2008 Summer  2008  Summer         Beijing  Athletics   
24879  2008 Summer  2008  Summer         Beijing  Athletics   
24880  2012 Summer  2012  Summer          London  Athletics   
24881  2012 Summer  2012  Summer          London  Athletics   
24882  2012 Summer  2012  Summer          London  Athletics   
24883  2016 Summer  2016  Summer  Rio de Janeiro  Athletics   
24884  2016 Summer  2016  Summer  Rio de Janeiro  Athletics   
24885  2016 Summer  2016  Summer  Rio de Janeiro  Athletics   

                                      Event Medal  
24876            Athletics Men's 200 metres   NaN  
24877            Athletics Men's 100 metres  Gold  
24878            Athletics Men's 200 metres  Gold  
24879  Athletics Men's 4 x 100 metres Relay   NaN  
24880            Athletics Men's 100 metres  Gold  
24881            Athletics Men's 200 metres  Gold  
24882  Athletics Men's 4 x 100 metres Relay  Gold  
24883            Athletics Men's 100 metres  Gold  
24884            Athletics Men's 200 metres  Gold  
24885  Athletics Men's 4 x 100 metres Relay  Gold

**Ответ** Без медали